src/store/products/selectors.js

```diff
@@ -0,0 +1,29 @@
+export const selectAllProducts = (state) => state.products;
+
+export const selectProductCategories = (state) => {
+  const categories = [];
+  state.products.forEach((product) => {
+    product.categories.forEach((category) => {
+      if (categories.indexOf(category) === -1) {
+        categories.push(category);
+      }
+    });
+  });
+  return categories;
+};
+
+export const selectProductById = (id) => (state) =>
+  state.products.find((p) => p.id === id);
+
+export const selectProductsByCart = (cart) => (state) => {
+  const products = [];
+  cart.forEach((product) =>
+    products.push(state.products.find((p) => p.id === product.productId))
+  );
```

Comment:
haha I'll get rid of all your `forEach`s, this we leave only to cause sideEffects like, doing an API call for each of them
This I would do:
```
const products = cart.reduce((acc, p) => {
    cont pToAdd = state.products.find((p) => p.id === product.productId);
    return [...acc, pToAdd] 
}, [])
// or you can one-line this with the find inside the return array
```

Reply:
aaa even easier and makes more sense!:
```
const products = cart.map(pro => state.products.find((p) => p.id === pro.productId))
```